qa/tasks/cephfs/test_volumes.py

```diff
@@ -7680,6 +7687,719 @@ def test_subvolume_snapshot_clone_with_no_wait_not_enabled(self):
+        Sometimes during testing of clone progress bars, and sometimes
+        otherwise too, an extra progress bar is seen with message "Global
+        Recovery Event".
+
+        This extra progress bar interferes with testing of progress bars for
+        cloning. This helper methods removes (filters out) this extra event.
+        '''
+        gre_key = None # key for global recovery event
+
+        for k, v in progress_events.items():
+            if 'global recovery event' in v['message'].lower():
+                gre_key = k
+                # Don't use "break". Usually there is only one "Global Recovery
+                # Event" in "progress_events" dictionary printed by "ceph
+                # status --format json-pretty". But let the loop run and catch
+                # and remove them in case there might be multiple of them here.
+                # Else, it'll interfere and cause tests to fail.
+                #break
+
+        if gre_key is not None:
+            progress_events.pop(gre_key)
+
+        return progress_events
+
+    def get_pevs_from_ceph_status(self, clones=None, check=True):
+        o = self.get_ceph_cmd_stdout('status --format json-pretty')
+        o = json.loads(o)
+
+        try:
+            pevs = o['progress_events'] # pevs = progress events
+        except KeyError as e:
+            try:
+                if check and clones:
+                    self.__check_clone_state('completed', clone=clones, timo=1)
+            except:
+                msg = ('Didn\'t find expected entries in dictionary '
+                       '"progress_events" which is obtained from the '
+                       'output of command "ceph status".\n'
+                       f'Exception - {e}\npev -\n{pevs}')
+                raise Exception(msg)
+
+        pevs = self.filter_out_gre_from_pevs(pevs)
+
+        return pevs
+
+    def test_when_clones_less_than_cloner_threads(self):
+        '''
+        Test that one progress bar is printed in output of "ceph status" output
+        when number of clone jobs is less than number of cloner threads.
+        '''
+        v = self.volname
+        sv = 'sv1'
+        ss = 'ss1'
+        # XXX: "clone" must be part of clone name for sake of tearDown()
+        c = 'ss1clone1'
+
+        self.run_ceph_cmd(f'fs subvolume create {v} {sv} --mode=777')
+        size = self._do_subvolume_io(sv, None, None, 10, 1024)
+
+        self.run_ceph_cmd(f'fs subvolume snapshot create {v} {sv} {ss}')
+        self.wait_till_rbytes_is_right(v, sv, size)
+
+        self.run_ceph_cmd(f'fs subvolume snapshot clone {v} {sv} {ss} {c}')
+
+        with safe_while(tries=10, sleep=1) as proceed:
+            while proceed():
+                pev = self.get_pevs_from_ceph_status(c)
+
+                if len(pev) < 1:
+                   time.sleep(1)
+                   continue
+                elif len(pev) > 1:
+                    raise RuntimeError('For 1 clone "ceph status" output has 2 '
```

Comment:
There's no progress event. I've run these tests many many times with teuth, it never happened.

Reply:
Have you checked if there are other users of progress module? Could it be the case that something kicks off a progress event that might cause this to fail. Could you share a sample json for this here so that we can assess if we can handle such spurious events that might show up?

Reply:
> Have you checked if there are other users of progress module Could it be the case that something kicks off a progress event that might cause this to fail.

Nope, it was never needed. I've run tests along with this version and it's earlier version to the point all tests were green every time. Individually it must have been 50 times. No event other than `global recovery event` are launched. So I am not worried about other progress events interfering without our tests.

>  Could you share a sample json for this here so that we can assess if we can handle such spurious events that might show up?

Sure -

```
    "progress_events": {
        "41df2eae-7750-4e74-90a3-90f7e9c69a19": {
            "message": "4 ongoing clones; Avg progress is 24.87% (14s)\n      [======......................] (remaining: 44s)",
            "progress": 0.24869999289512634,
            "add_to_ceph_s": true
        },
        "6f4b5599-6a30-41dd-8dba-837154cedeb7": {
            "message": "6 ongoing+pending clones; Avg progress is 22.222% (5s)\n      [======......................] (remaining: 17s)",
            "progress": 0.22222000360488892,
            "add_to_ceph_s": true
        }
```

Reply:
Thanks for the information. Now, thinking again, is there a need to have randomised keys for events? Cannot these be static keys (something like `mgr-volumes-ongoing-clones` and `mgr-volumes-total-clones`)?

Reply:
> is there a need to have randomised keys for events?

Don't remember encountering use of randomized keys during this work.

> Cannot these be static keys (something like mgr-volumes-ongoing-clones and mgr-volumes-total-clones)?

Sounds better to me.

Reply:
Done.

Reply:
https://github.com/ceph/ceph/pull/54620#issuecomment-2181308240